docs/architecture/adr-076-sign-mode-unified.md

```diff
@@ -0,0 +1,349 @@
+
+This metadata does not need to be stored on the signer app since, by using a chunked and merkleized version of the metadata, chunks representing the data types involved in the messages can be delivered on-demand to the signer, and the root hash of the tree can be computed from those chunks and proofs from the remaining chunks, thus verifying the metadata integrity. On the signer app side, what the user ends up signing is the `metadata root hash + blob`, this means that the signed transaction can be verified by the validator by checking that the metadata root hash in the signed transaction matches the one locally stored in the node.
+
+## Context
+
+## Challenges with Current sign modes
+
+1. **Complex codebase:**
+   Maintaining both sign modes increases the maintenance overhead and the complexity of the codebase.
+
+2. **Inconsistent signing processes across different Cosmos chains:**
+   Not all Cosmos chains use the same sign mode, which creates inconsistencies in the ecosystem.
+
+3. **Blind signing:**
+   When wallets cannot decode a transaction or smart contract call, it leads to blind signing. Blind signing exposes users to potential security risks, as they may unknowingly sign malicious transactions.
+
+4. **Cross-language Types Inconsistency:**
+   Implementing renderers for each language in which a signer would be implemented can lead to inconsistencies in type interpretation.
+
+5. **Upgradeability:**
+   When new types are added to the system, wallets need to synchronize their renderers to match these new types. This process can lead to delays in supporting new types of transactions.
+
+
+## Main components
+
+Key components of `SIGN_MODE_UNIFIED` include:
+
+* *Metadata struct*: A JSON file containing the definitions of all data types for each message of each module, down to primitive types.
+* *Metadata Digest*: A compact representation of the complete metadata set used for transaction construction. When hashed (digest hash), this struct serves as the root of the merkle tree, enabling efficient verification of metadata integrity.
+* *Merkleized Structure*: The metadata is organized into a merkle tree, allowing for efficient proofs and verification.
+
+### Metadata file
+
+The metadata file is a JSON file that contains all the types and messages definitions used in the SDK, is meant to be generated at build time and stored in the node, using the protobuf definition files as source. This file (in fact, chunks of its encoded version) are sent to the signer in order to be used to decode and show the transaction's fields without the need to store locally all definitions of types and messages. In the end, all signers should only be compliant with the primitive types and their string representations. This ensures consistency and reliability in type interpretation
```

Comment:
Does this mean that there is a new "handshake" from the signer to the node they are broadcasting to?

They now much request the relevant metadata for their transaction?

If so, does this mean anything for offline signing?

Reply:
The metadata chunks are requested to the node on-demand lets say, by the signer.

> They now much request the relevant metadata for their transaction?
Yes, the signer asks for the chunks involved in the tx and the proofs to rebuild the tree hash

> If so, does this mean anything for offline signing?
For offline signing, one option is that the wallet should store locally the full metadata file, so the entire signing and metadata verification happens offline. The "downside" is that this metadata should be updated every time it changes (a simple call to a node will be enough IMO). If the signer uses an outdated metadata to sign, worst case scenario is that the tx will be rejected at validation time.